File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax/UsingStatementSyntax.cs
using Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
{
	public sealed class UsingStatementSyntax : StatementSyntax
	{
		internal ExpressionSyntax _expression;

		internal SyntaxNode _variables;

		public SyntaxToken UsingKeyword => new SyntaxToken(this, ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.UsingStatementSyntax)base.Green)._usingKeyword, base.Position, 0);

		public ExpressionSyntax Expression => GetRed(ref _expression, 1);

		public SeparatedSyntaxList<VariableDeclaratorSyntax> Variables
		{
			get
			{
				SyntaxNode red = GetRed(ref _variables, 2);
				return (red == null) ? default(SeparatedSyntaxList<VariableDeclaratorSyntax>) : new SeparatedSyntaxList<VariableDeclaratorSyntax>(red, GetChildIndex(2));
			}
		}

		internal UsingStatementSyntax(GreenNode green, SyntaxNode parent, int startLocation)
			: base(green, parent, startLocation)
		{
		}

		internal UsingStatementSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, KeywordSyntax usingKeyword, ExpressionSyntax expression, SyntaxNode variables)
			: this(new Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.UsingStatementSyntax(kind, errors, annotations, usingKeyword, (expression != null) ? ((Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax)expression.Green) : null, variables?.Green), null, 0)
		{
		}

		public UsingStatementSyntax WithUsingKeyword(SyntaxToken usingKeyword)
		{
			return Update(usingKeyword, Expression, Variables);
		}

		public UsingStatementSyntax WithExpression(ExpressionSyntax expression)
		{
			return Update(UsingKeyword, expression, Variables);
		}

		public UsingStatementSyntax WithVariables(SeparatedSyntaxList<VariableDeclaratorSyntax> variables)
		{
			return Update(UsingKeyword, Expression, variables);
		}

		public UsingStatementSyntax AddVariables(params VariableDeclaratorSyntax[] items)
		{
			return WithVariables(Variables.AddRange(items));
		}

		internal override SyntaxNode GetCachedSlot(int i)
		{
			return i switch
			{
				1 => _expression, 
				2 => _variables, 
				_ => null, 
			};
		}

		internal override SyntaxNode GetNodeSlot(int i)
		{
			return i switch
			{
				1 => Expression, 
				2 => GetRed(ref _variables, 2), 
				_ => null, 
			};
		}

		public override TResult Accept<TResult>(VisualBasicSyntaxVisitor<TResult> visitor)
		{
			return visitor.VisitUsingStatement(this);
		}

		public override void Accept(VisualBasicSyntaxVisitor visitor)
		{
			visitor.VisitUsingStatement(this);
		}

		public UsingStatementSyntax Update(SyntaxToken usingKeyword, ExpressionSyntax expression, SeparatedSyntaxList<VariableDeclaratorSyntax> variables)
		{
			if (usingKeyword != UsingKeyword || expression != Expression || variables != Variables)
			{
				UsingStatementSyntax usingStatementSyntax = SyntaxFactory.UsingStatement(usingKeyword, expression, variables);
				SyntaxAnnotation[] annotations = GetAnnotations();
				if (annotations != null && annotations.Length > 0)
				{
					return SyntaxNodeExtensions.WithAnnotations(usingStatementSyntax, annotations);
				}
				return usingStatementSyntax;
			}
			return this;
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/UsingStatementSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
Imports System

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax
	Public NotInheritable Class UsingStatementSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.StatementSyntax
		Friend _expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax

		Friend _variables As SyntaxNode

		Public ReadOnly Property Expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax
			Get
				Return MyBase.GetRed(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax)(Me._expression, 1)
			End Get
		End Property

		Public ReadOnly Property UsingKeyword As Microsoft.CodeAnalysis.SyntaxToken
			Get
				Return New Microsoft.CodeAnalysis.SyntaxToken(Me, DirectCast(MyBase.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.UsingStatementSyntax)._usingKeyword, MyBase.Position, 0)
			End Get
		End Property

		Public ReadOnly Property Variables As SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax)
			Get
				Dim variableDeclaratorSyntaxes As SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax)
				Dim red As SyntaxNode = MyBase.GetRed(Me._variables, 2)
				variableDeclaratorSyntaxes = If(red Is Nothing, New SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax)(), New SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax)(red, MyBase.GetChildIndex(2)))
				Return variableDeclaratorSyntaxes
			End Get
		End Property

		Friend Sub New(ByVal green As GreenNode, ByVal parent As SyntaxNode, ByVal startLocation As Integer)
			MyBase.New(green, parent, startLocation)
		End Sub

		Friend Sub New(ByVal kind As Microsoft.CodeAnalysis.VisualBasic.SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal usingKeyword As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.KeywordSyntax, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax, ByVal variables As SyntaxNode)
			MyClass.New(New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.UsingStatementSyntax(kind, errors, annotations, usingKeyword, If(expression IsNot Nothing, DirectCast(expression.Green, Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.ExpressionSyntax), Nothing), If(variables IsNot Nothing, variables.Green, Nothing)), Nothing, 0)
		End Sub

		Public Overrides Function Accept(Of TResult)(ByVal visitor As VisualBasicSyntaxVisitor(Of TResult)) As TResult
			Return visitor.VisitUsingStatement(Me)
		End Function

		Public Overrides Sub Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.VisualBasicSyntaxVisitor)
			visitor.VisitUsingStatement(Me)
		End Sub

		Public Function AddVariables(ByVal ParamArray items As Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax()) As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax
			Return Me.WithVariables(Me.Variables.AddRange(items))
		End Function

		Friend Overrides Function GetCachedSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.SyntaxNode
			Dim syntaxNode As Microsoft.CodeAnalysis.SyntaxNode
			Dim num As Integer = i
			If (num = 1) Then
				syntaxNode = Me._expression
			ElseIf (num = 2) Then
				syntaxNode = Me._variables
			Else
				syntaxNode = Nothing
			End If
			Return syntaxNode
		End Function

		Friend Overrides Function GetNodeSlot(ByVal i As Integer) As SyntaxNode
			Dim expression As SyntaxNode
			Dim num As Integer = i
			If (num = 1) Then
				expression = Me.Expression
			ElseIf (num = 2) Then
				expression = MyBase.GetRed(Me._variables, 2)
			Else
				expression = Nothing
			End If
			Return expression
		End Function

		Public Function Update(ByVal usingKeyword As Microsoft.CodeAnalysis.SyntaxToken, ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax, ByVal variables As SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax
			Dim usingStatementSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax
			If (usingKeyword <> Me.UsingKeyword OrElse expression <> Me.Expression OrElse variables <> Me.Variables) Then
				Dim usingStatementSyntax1 As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax = Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.UsingStatement(usingKeyword, expression, variables)
				Dim annotations As SyntaxAnnotation() = MyBase.GetAnnotations()
				usingStatementSyntax = If(annotations Is Nothing OrElse CInt(annotations.Length) <= 0, usingStatementSyntax1, usingStatementSyntax1.WithAnnotations(annotations))
			Else
				usingStatementSyntax = Me
			End If
			Return usingStatementSyntax
		End Function

		Public Function WithExpression(ByVal expression As Microsoft.CodeAnalysis.VisualBasic.Syntax.ExpressionSyntax) As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax
			Return Me.Update(Me.UsingKeyword, expression, Me.Variables)
		End Function

		Public Function WithUsingKeyword(ByVal usingKeyword As Microsoft.CodeAnalysis.SyntaxToken) As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax
			Return Me.Update(usingKeyword, Me.Expression, Me.Variables)
		End Function

		Public Function WithVariables(ByVal variables As SeparatedSyntaxList(Of Microsoft.CodeAnalysis.VisualBasic.Syntax.VariableDeclaratorSyntax)) As Microsoft.CodeAnalysis.VisualBasic.Syntax.UsingStatementSyntax
			Return Me.Update(Me.UsingKeyword, Me.Expression, variables)
		End Function
	End Class
End Namespace